CONFIGURATION.md
# Coordinator

As of right now all config options for the coordinator are specified via environment variables. In a docker compose file
that corresponds to the `environment` section. The following options are available right now:

|                      Name | Default value | Description                                                                                                                                                                                                                        |
|--------------------------:|:-------------:|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|          ``MAX_BUILDERS`` |       1       | The amount of package builder that are allowed to run in parallel. Be aware setting this higher than one might quickly starve the system of resource as compiling source code is intensive work.                                   |
|           ``MAX_RETRIES`` |       3       | After a build failed, the build will be restarted after five minutes. Should this happen more than the specified value, the build will 'time out' until the next update check interval when it will give it a set of new attempts. |
|              ``MAX_LOGS`` |       5       | How many logs to keep from failed builds before old ones get deleted. If set to 0, they never get deleted.                                                                                                                         |
| ``UPDATE_CHECK_INTERVAL`` |      240      | The amount of time, in minutes, between querying the AUR for package updates.                                                                                                                                                      |
|                  ``PORT`` |     3200      | The port the web server should bind to.                                                                                                                                                                                            |
|         ``BUILDER_IMAGE`` |  aur_worker   | The docker tag of the image for the builder unit.                                                                                                                                                                                  |
|             ``REPO_NAME`` |      aur      | The name for the repository that the packages will be placed in.                                                                                                                                                                   |
|          ``MEMORY_LIMIT`` |               | If set, limits the maximum amount of RAM a worker is allowed to use in bytes. Max usage of all workers should be MAX_BUILDER * MEMORY_LIMT.                                                                                        |

# Archie

Archie stores it's configuration files in the users home directory under `$HOME/.config/archie/config.toml`. Alternative
profile use the name as the profile instead of `config` for their file name. Right now all options that can be set can
be configured by just running `archie init`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Config option for update check interval
- Users can request rebuilds manually
- The memory usage of the worker can now be limited
- Packages can be added from a git repositories url
- A ssh key gets created a startup to allow access to git repositories with access control
- Logs now be inspected using the cli tool

### Changed

- Makepkg is now being used instead of paru
- Logs are no longer being output to the coordinator logs. They can now be retrieved with the cli tool or read from the
  `/logs` directory withing the container
    - Added a `MAX_LOGS` environment variable. Logs get rotated if they exceed the maximum number

### Fixed

- Database was not created if no packages were being tracked
- Old database file not being deleted if the name of the repository is changed
- Coordinator not shutting down if any part of it shuts down unexpectedly

## [0.2.0] - 2024-12-12

### Added

- Discovery of dependencies
- More info to the status command, including server info, warnings, and the list of packages
- Multiple configuration profiles for the client
- Package remove functionality
- Apps print their version

### Changed

- Set the worker container's name to the package name
- Default number of builders changed to 1
- Changed error levels to warnings or debug for better clarity
- Refactored internal communication mechanisms for better maintainability

### Fixed

- Fixed non-built packages not being built upon restart
- Fixed orchestrator reporting forcefully stopped containers as exited abnormally
- Fixed repository trying to remove never-built packages
- Fixed scheduler bugs:
    - Prevent unnecessary waiting during the first loop
    - React to new messages immediately
- Fixed worker bugs:
    - Corrected endpoint configuration
    - Improved log levels, moving less critical items to debug
- Fixed repository bugs:
    - Prevent running commands when no repository exists
    - Corrected error reporting for repository operations
- Fixed max retries not being respected
- Fixed bugs with CLI and webserver endpoints:
    - Corrected struct usage in remove command
    - Fixed swapped webserver endpoints

### Removed

- Removed unneeded dependencies
- Removed unneeded log statements
- Removed unused repository lock field
- Removed unneeded imports
- Removed accidental println statements

## [0.1.0] - 2024-12-06

Initial implementation of the application

- Core functionality to build and rebuild packages as necessary when updates are pushed
- CLI application for sending build requests
- Configuration management through environment variables
- Graceful termination of the application using signal handling

[Unreleased]: https://git.techmayhem.net/techMayhem/archie/compare/v0.2.0...HEAD

[0.2.0]: https://git.techmayhem.net/techMayhem/archie/compare/v0.1.0...v0.2.0

[0.1.0]: https://git.techmayhem.net/techMayhem/archie/releases/tag/v0.1.0

src/coordinator/config.rs
use coordinator::{env_or, env_or_none};
use std::sync::LazyLock;
use tracing::info;

static CONFIG: LazyLock<Config> = LazyLock::new(load);

#[derive(Debug)]
struct Config {
    max_builders: usize,
    max_retries: u8,
    max_logs: u8,
    update_check_interval: u32,
    port: u32,
    image: String,
    repo_name: String,
    memory_limit: Option<i64>,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            max_builders: 1,
            max_retries: 3,
            max_logs: 5,
            update_check_interval: 240,
            port: 3200,
            image: "aur_worker".to_string(),
            repo_name: "aur".to_string(),
            memory_limit: None,
        }
    }
}

fn load() -> Config {
    let config = load_from_env();
    info!("Loaded config: {config:#?}");
    config
}

fn load_from_env() -> Config {
    let default = Config::default();

    Config {
        max_builders: env_or("MAX_BUILDERS", default.max_builders),
        max_retries: env_or("MAX_RETRIES", default.max_retries),
        max_logs: env_or("MAX_LOGS", default.max_retries),
        update_check_interval: env_or("UPDATE_CHECK_INTERVAL", default.update_check_interval),
        port: env_or("PORT", default.port),
        image: env_or("BUILDER_IMAGE", default.image),
        repo_name: env_or("REPO_NAME", default.repo_name),
        memory_limit: env_or_none("MEMORY_LIMIT"),
    }
}

pub fn max_builders() -> usize {
    CONFIG.max_builders
}

pub fn max_retries() -> u8 {
    CONFIG.max_retries
}

pub fn max_logs() -> u8 { CONFIG.max_logs }

pub fn update_check_interval() -> u32 {
    CONFIG.update_check_interval
}

pub fn port() -> u32 {
    CONFIG.port
}

pub fn image() -> String {
    CONFIG.image.clone()
}

pub fn repo_name() -> String {
    CONFIG.repo_name.clone()
}

pub fn max_memory() -> Option<i64> { CONFIG.memory_limit }